internal/profiling/profiling.go

```diff
@@ -0,0 +1,213 @@
+// whether it's a client stat or a server stat; so you should be able to filter
+// for the right type of stats in post-processing.
+// SetEnabled is the internal
+func SetEnabled(enabled bool) {
+	if enabled {
+		atomic.StoreUint32(&profilingEnabled, 1)
+	} else {
+		atomic.StoreUint32(&profilingEnabled, 0)
+	}
+}
+
+// A Timer represents the wall-clock beginning and ending of a logical
+// operation.
+type Timer struct {
+	TimerTag string
+	Begin    time.Time
+	End      time.Time
+	GoId     int64
+}
+
+// NewTimer creates and returns a new Timer object. This is useful when you
+// don't already have a Stat object to associate this Timer with; for example,
+// before the context of a new RPC query is created, a Timer may be needed to
+// measure transport-related operations.
+//
+// Use AppendTimer to append the returned Timer to a Stat.
+func NewTimer(timerTag string) *Timer {
+	return &Timer{
+		TimerTag: timerTag,
+		Begin:    time.Now(),
+		GoId:     goId(),
+	}
+}
+
+// A Stat is a collection of Timers that represent timing information for
+// different components within this Stat. For example, a Stat may be used to
+// reference the entire lifetime of an RPC request, with Timers within it
+// representing different components such as encoding, compression, and
+// transport.
+//
+// The user is expected to use the included helper functions to do operations
+// on the Stat such as creating or appending a new timer. Direct operations on
+// the Stat's exported fields (which are exported for encoding reasons) may
+// lead to data races.
+type Stat struct {
+	StatTag  string
+	Timers   []Timer
+	Metadata []byte
+	mu       sync.Mutex
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation.
+const defaultStatAllocatedTimers int32 = 128
+
+// NewStat creates and returns a new Stat object.
+func NewStat(statTag string) *Stat {
+	return &Stat{
+		StatTag: statTag,
+		Timers:  make([]Timer, 0, defaultStatAllocatedTimers),
+	}
+}
+
+// NewTimer creates a Timer object within the given stat if stat is non-nil.
+// The value passed in timerTag will be attached to the newly created Timer.
+// NewTimer also automatically sets the Begin value of the Timer to the current
+// time.
+//
+// The user is expected to call stat.Egress with the returned index as argument
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) (index int) {
```

Comment:
Named return value parameters don't lead to very readable code. Do you have a strong reason to use them here?

Excerpt from internal Go readability doc:
```
Don't name result parameters just to avoid declaring a var inside the function; that trades off a minor implementation brevity at the cost of unnecessary API verbosity.

Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

Finally, in some cases you need to name a result parameter in order to change it in a deferred closure. That is always OK.
```

Reply:
Not really, I don't have a real reason. I agree that it's much more readable to declare the variables inside the function.

Done.